rosetta/retriever/retriever.go

```diff
@@ -131,35 +91,64 @@ func (r *Retriever) Current() (identifier.Block, time.Time, error) {
 	return block, header.Timestamp, nil
 }
 
-func (r *Retriever) Balances(network identifier.Network, block identifier.Block, account identifier.Account, currencies []identifier.Currency) ([]object.Amount, error) {
+func (r *Retriever) Balances(block identifier.Block, account identifier.Account, currencies []identifier.Currency) (identifier.Block, []rosetta.Amount, error) {
+
+	// Run validation on the block ID. This also fills in missing information.
+	completeBlock, err := r.validate.Block(block)
+	if err != nil {
+		return identifier.Block{}, nil, fmt.Errorf("could not validate block: %w", err)
+	}
+
+	// Run validation on the account ID. This uses the chain ID to check the
+	// address validation.
+	err = r.validate.Account(account)
+	if err != nil {
+		return identifier.Block{}, nil, fmt.Errorf("could not validate account: %w", err)
+	}
+
+	// Run validation on the currencies. This checks basically if we know the
+	// currency and if it has the correct decimals set, if they are set.
+	for index, currency := range currencies {
+		completeCurrency, err := r.validate.Currency(currency)
+		if err != nil {
+			return identifier.Block{}, nil, fmt.Errorf("could not validate currency: %w", err)
+		}
+		currencies[index] = completeCurrency
+	}
 
 	// get the cadence value that is the result of the script execution
-	amounts := make([]object.Amount, 0, len(currencies))
+	amounts := make([]rosetta.Amount, 0, len(currencies))
 	address := cadence.NewAddress(flow.HexToAddress(account.Address))
 	for _, currency := range currencies {
 		getBalance, err := r.generator.GetBalance(currency.Symbol)
 		if err != nil {
-			return nil, fmt.Errorf("could not generate script: %w", err)
+			return identifier.Block{}, nil, fmt.Errorf("could not generate script: %w", err)
 		}
 		value, err := r.invoke.Script(block.Index, getBalance, []cadence.Value{address})
 		if err != nil {
-			return nil, fmt.Errorf("could not invoke script: %w", err)
+			return identifier.Block{}, nil, fmt.Errorf("could not invoke script: %w", err)
 		}
 		balance, ok := value.ToGoValue().(uint64)
 		if !ok {
-			return nil, fmt.Errorf("could not convert balance (type: %T)", value.ToGoValue())
+			return identifier.Block{}, nil, fmt.Errorf("could not convert balance (type: %T)", value.ToGoValue())
 		}
-		amount := object.Amount{
+		amount := rosetta.Amount{
 			Currency: currency,
 			Value:    strconv.FormatUint(balance, 10),
 		}
 		amounts = append(amounts, amount)
 	}
 
-	return amounts, nil
+	return completeBlock, amounts, nil
 }
 
-func (r *Retriever) Block(network identifier.Network, id identifier.Block) (*object.Block, []identifier.Transaction, error) {
+func (r *Retriever) Block(id identifier.Block) (*rosetta.Block, []identifier.Transaction, error) {
+
+	// Run validation on the block ID. This also fills in missing information.
```

Comment:
What missing information? It doesn't seem like the responsibility of a validator to edit what it validates.

Reply:
I disagree; validating an identifier can also mean that it completes it to the full identifier. The issue is with the Rosetta specification, which specifies a single `block identifier` that has both a `hash` and an `index` field. When returning values, it needs to specify both, but when being requested, either one can be omitted. This means that we need to fill in the empty field so it's correct when returned.

Reply:
I better design, imo, would have been for Rosetta to handle different types of unique identifiers separately, like we do (height/hash, maybe others like commit/view).